sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -676,6 +678,43 @@ class KeyValueGroupedDataset[K, V] private[sql](
     )
   }
 
+  private[sql] def transformWithState[U: Encoder](
+       statefulProcessor: StatefulProcessor[K, V, U],
+       timeoutMode: TimeoutMode,
+       eventTimeColumnName: String,
+       outputMode: OutputMode): Dataset[U] = {
+    val existingWatermarkDelay = logicalPlan.flatMap {
+      case EventTimeWatermark(_, delay, _) => Seq(delay)
+      case _ => Seq()
+    }
+
+    if (existingWatermarkDelay.isEmpty) {
+      throw QueryCompilationErrors.cannotAssignEventTimeColumn()
+    }
+
+    val transformWithState = TransformWithState[K, V, U](
+      groupingAttributes,
+      dataAttributes,
+      statefulProcessor,
+      timeoutMode,
+      outputMode,
+      child = logicalPlan
+    )
+
+    val twsDS = Dataset[U](
+      sparkSession,
+      transformWithState
+    )
+
+    val delay = existingWatermarkDelay.head
+
+    Dataset[U](sparkSession, EliminateEventTimeWatermark(
+      UpdateEventTimeWatermarkColumn(
```

Comment:
How does this modify the plan exactly ? if we have 2 or more `TransformWithState` operators chained together, how would this work exactly ?

Reply:
This will add a new node `UpdateEventTimeColumn` after the `TransformWithState` node. The `UpdateEventTimeColumn` will update the column metadata and tag the new column as eventTime. 

Same thing will happen if we have 2 or more `TransformWithState` operators chained together. After each TWS operator, we will add the `UpdateEventTimeColumn` node to tag the correct event time column.